examples/print-dialog/print_dialog.py

```diff
@@ -0,0 +1,540 @@
+printer_type_selection_menu["values"] = state.available_printer_types
+printer_type_selection_menu.grid(row=0, column=1, sticky="w")
+
+printer_selection_label = tk.Label(frame, text="Printer: ")
+printer_selection_label.grid(row=1, column=0, sticky="w")
+
+printer_selection_menu = ttk.Combobox(frame, width=34, state="readonly", textvariable=selected_printer)
+printer_selection_menu["values"] = state.printer_selection_menu_options
+printer_selection_menu.grid(row=1, column=1, sticky="w")
+
+print_setting_selection_label = tk.Label(frame, text="Print Setting: ")
+print_setting_selection_label.grid(row=2, column=0, sticky="w")
+
+print_setting_selection_menu = ttk.Combobox(frame, width=34, state="readonly", textvariable=selected_print_setting)
+print_setting_selection_menu["values"] = state.print_setting_selection_menu_options
+print_setting_selection_menu.grid(row=2, column=1, sticky="w")
+
+job_name_label = tk.Label(frame, text="Job Name: ")
+job_name_label.grid(row=3, column=0, sticky="w")
+job_name_entry = tk.Entry(frame, textvariable=job_name)
+job_name_entry.grid(row=3, column=1, sticky="w")
+
+print_buttons_frame = tk.Frame(frame)
+print_buttons_frame.grid(row=4, column=1, sticky="w")
+queue_print_button = tk.Button(
+    print_buttons_frame,
+    text="Upload to Queue",
+    command=lambda: upload_print(state),
+)
+queue_print_button.pack(side=tk.LEFT)
+
+print_now_button = tk.Button(
+    print_buttons_frame,
+    text="Print Now",
+    command=lambda: upload_print(state, print_now=True),
+)
+print_now_button.pack(side=tk.LEFT)
+
+bottom_frame = tk.Frame(root)
+bottom_frame.pack(anchor="w", padx=10, pady=10)
+
+progress_label = tk.Label(bottom_frame, textvariable=progress_label_textvar)
+progress_label.grid(row=0, column=0, sticky="w")
+progress_bar = ttk.Progressbar(
+    bottom_frame,
+    orient="horizontal",
+    length=200,
+    mode="determinate",
+    variable=progress_bar_value,
+)
+progress_bar.grid(row=1, column=0, sticky="w")
+progress_bar.grid_remove()  # Hide it initially
+
+
+################################# Setup event handling after UI elements exist:
+def on_selected_printer_type_change(state: AppState, selected_printer_type: tk.StringVar) -> None:
+    state.set_selected_printer_type(selected_printer_type.get())
+    sync_printer_dropdown_with_state(state)
+    sync_print_setting_dropdown_with_state(state)
+    sync_printing_button_options(state)
+
+
+def on_selected_printer_change(state: AppState, selected_printer: tk.StringVar) -> None:
+    state.selected_printer = selected_printer.get()
+    sync_printing_button_options(state)
+
+
+def on_selected_print_setting_change(state: AppState, selected_print_setting: tk.StringVar) -> None:
+    state.selected_print_setting = selected_print_setting.get()
+
+
+def on_job_name_change(state: AppState, job_name: tk.StringVar) -> None:
+    state.job_name = job_name.get()
```

Comment:
I'm not too familiar with TkInter, but I wonder if this stuff would be cleaner as a single idempotent update() call which can be free spammed and scheduled any time state updates.

Reply:
I thought about that as well but Tkinter is structured around individual GUI element callbacks containing event data so I preferred to stay in line with that for this example.